### PR TITLE
[GH-25484] Fix draft removal on post deletion

### PR DESCRIPTION
**Pull request by Utsav-Ladani**
#### Summary
- This PR fixes the issue with the draft removal on deleting the post
- See this issue for more info: https://github.com/mattermost/mattermost/issues/25484

#### Steps to reproduce and test
- Create a post from account `A`
- Write something as a draft on that post from a different account `B`
- Delete a post from account `A`
- See Draft section on account `B`
- In the older version, the draft was still visible in the `Draft` section on the right side. But in this fix, it will be removed.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost/issues/25484
https://mattermost.atlassian.net/browse/MM-55613

#### Release Note

```release-note
Fix Draft removal after deleting the post
```

### CODE REVIEW

**Comment by Utsav-Ladani** [comment]
### Why is it a Draft PR?

There are two issues we have to resolve. I need some help on that:
1. It works fine for newly created drafts, but older ones persist in local storage. We must write a database script to delete a draft whose parent post is deleted. Run this script only once when the product version is updated. I need to figure out where to put it.
2. I cannot figure out why the `removeGlobalItem` function doesn't reset the draft message in the text editor alone. I have to use the `setGlobalItem` function to set it empty and then use the `removeGlobalItem` function to remove it as expected. I need help to figure out why this is happening.

Thanks

CC: @devinbinnie

**Comment by devinbinnie** [comment]
Thanks for the PR @Utsav-Ladani! As for your questions...

> 1. It works fine for newly created drafts, but older ones persist in local storage. We must write a database script to delete a draft whose parent post is deleted. Run this script only once when the product version is updated. I need to figure out where to put it.

So we have a folder in the `server` part of the repo that holds migrations. It's `/server/channels/db/migrations` and you can create migrations for both MySQL and Postgres here. You can run `make new-migration name=<>` to generate everything for you.

> 2. I cannot figure out why the `removeGlobalItem` function doesn't reset the draft message in the text editor alone. I have to use the `setGlobalItem` function to set it empty and then use the `removeGlobalItem` function to remove it as expected. I need help to figure out why this is happening.

cc @hmhealey?

**Comment by Utsav-Ladani** [comment]
Thanks for the help @devinbinnie 

I added scripts for database migration. Please let me know if there is something wrong

File: server/channels/db/migrations/mysql/000119_remove_orphan_draft.up.sql [hunk 1]
```diff
@@ -0,0 +1,15 @@
+-- Remove drafts that are not referenced by any other post.
+DELETE FROM drafts
```
Comment (devinbinnie) [inline server/channels/db/migrations/mysql/000119_remove_orphan_draft.up.sql:2]: @Utsav-Ladani Sorry I didn't think this through, this actually should a batch migration. We don't want to lock the table during migration as that could cause bad behaviour for users.

Instead, we can do a batch migration as seen here (we've already done this for drafts for a different reason :P)
https://github.com/mattermost/mattermost/blob/master/server/channels/jobs/delete_empty_drafts_migration/delete_empty_drafts_migration.go

Comment (Utsav-Ladani) [reply server/channels/db/migrations/mysql/000119_remove_orphan_draft.up.sql]: Oh, I forgot about that thing. Now, I fixed it in the latest commit. Please let me know if I still need to correct it.

**Comment by Utsav-Ladani** [comment]
@devinbinnie I fixed the migration process and added a batch migration. Please let me know your suggestions.

**Comment by devinbinnie** [comment]
@Utsav-Ladani Thanks for updating. Looks like some of the checks are failing, can you have a look?

**Comment by hmhealey** [comment]
> > 2. I cannot figure out why the `removeGlobalItem` function doesn't reset the draft message in the text editor alone. I have to use the `setGlobalItem` function to set it empty and then use the `removeGlobalItem` function to remove it as expected. I need help to figure out why this is happening.
> 
> cc @hmhealey?

I think that's probably happening because of the logic where changes to a draft are reflected on the DOM. We don't want them fully connected (like `<textarea value={draft.message}/>`) because that causes typing to get slowed down a lot whenever JS lags. 

For this specific case, I'd guess that the problem logic is the `AdvancedCreatePost` and `AdvancedCreateComment` components' `getDerivedStateFromProps` static methods where we check if the draft has changed before modifying `updatedState`. To determine if that's the location of the problem, you could have that always modify the state in the draft and see if it fixes the issue. That'll break other things though, so it wouldn't be a permanent fix.

Alternatively, it might be an issue with `QuickInput` since it does something similar with `componentDidUpdate` to prevent passing the value directly into the textbox

All that said, I don't think it's that big of a deal if we empty the draft before removing it on the client. That code is pretty awkward to work with, so I wouldn't be surprised if that is just another quirk of the system that we need to work around

**Comment by hmhealey** [comment]
Those notes are perfect, thanks! Since this system is complicated, I really appreciate the reminder on how it works.

I didn't realize that deleting the draft set it to "local". That makes sense to me why it would screw up the logic in the `getDerivedStateFromProps` for those components. I'm hesitant to change that more since that logic was tricky to make work right in other cases, so I'm good with the current solution of updating the draft and then removing it

**Comment by Utsav-Ladani** [comment]
Thanks @hmhealey

**Comment by Utsav-Ladani** [comment]
Hello @devinbinnie, 

All PR checks are passed except `Enterprise CI/tests`. I try to see, but I am getting **404** on that check's [Details](https://github.com/mattermost/enterprise/actions/runs/7426018019) page.

Can you please help me to resolve this?

**Comment by Utsav-Ladani** [comment]
@devinbinnie No problem. Thanks for suggestion.

I remove all changes of `en.json`. Let me know is server/CI test is still failing.

**Review by devinbinnie — APPROVED** [review]
Thanks @Utsav-Ladani, this LGTM!

**Review by hmhealey — APPROVED** [review]
Thanks for the PR, and thanks again for sorting this out given the drafts system is so complicated.

I just fought with this same sort of issue for someone and ended up writing a one-off script to solve it, so it'll be nice to have this in. I'm not sure if this would've prevented that exact issue because it was caused by messing with the database directly, but it's good to know that we have a pattern for solving this now.

**Review by yasserfaraazkhan — APPROVED** [review]

**Comment by yasserfaraazkhan** [comment]
Tested on desktop mac and win 5.6

**Pull request merged**